### PR TITLE
feat: add auxiliary deposit functions

### LINKED ISSUE #375: Deposit refactoring in `phase0`
There are some additional helpers added in the `consensus-specs` for deposit handling (like `apply_deposit`).

Would be nice to sync this repo w/ those updates

https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md#deposits

Some missing functions we could add:

`add_validator_to_registry`
`apply_deposit`

### PR DESCRIPTION
**Pull request by jacobkaufmann**
resolves #375

### CODE REVIEW

**Comment by jacobkaufmann** [comment]
can remove generated `electra` spec if desired

File: ethereum-consensus/src/phase0/block_processing.rs [hunk 1]
```diff
@@ -311,6 +317,100 @@ pub fn get_validator_from_deposit(deposit: &Deposit, context: &Context) -> Valid
...
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    public_key: &BlsPublicKey,
+    withdrawal_credentials: &Bytes32,
+    amount: u64,
```
Comment (ralexstokes) [inline ethereum-consensus/src/phase0/block_processing.rs:376]: imo spec should say `Gwei` here, and we can do this here either way for a dash of type-safety

File: ethereum-consensus/src/phase0/block_processing.rs [hunk 2]
```diff
@@ -311,6 +317,100 @@ pub fn get_validator_from_deposit(deposit: &Deposit, context: &Context) -> Valid
...
+        ETH1_DATA_VOTES_BOUND,
+        VALIDATOR_REGISTRY_LIMIT,
+        EPOCHS_PER_HISTORICAL_VECTOR,
+        EPOCHS_PER_SLASHINGS_VECTOR,
+        MAX_VALIDATORS_PER_COMMITTEE,
+        PENDING_ATTESTATIONS_BOUND,
+    >,
+    public_key: BlsPublicKey,
+    withdrawal_credentials: Bytes32,
+    amount: u64,
```
Comment (ralexstokes) [inline ethereum-consensus/src/phase0/block_processing.rs:342]: update to `Gwei` if you change below

File: ethereum-consensus/src/altair/block_processing.rs [hunk 3]
```diff
@@ -159,6 +156,103 @@ pub fn process_attestation<
...
+        public_key.clone(),
+        withdrawal_credentials.clone(),
+        amount,
+        context,
+    );
+
+    Ok(())
+}
+
 pub fn process_deposit<
```
Comment (ralexstokes) [inline ethereum-consensus/src/altair/block_processing.rs:256]: I think we should also be able to delete this `process_deposit` definition and let the `spec-gen` handle it

File: ethereum-consensus/src/phase0/block_processing.rs [hunk 4]
```diff
@@ -311,6 +317,100 @@ pub fn get_validator_from_deposit(deposit: &Deposit, context: &Context) -> Valid
...
+    let signing_root = compute_signing_root(&mut deposit_message, domain)?;
+    if verify_signature(public_key, signing_root.as_ref(), signature).is_err() {
+        // NOTE: explicitly return with no error and also no further mutations to `state`
+        return Ok(());
+    }
+
+    add_validator_to_registry(
+        state,
+        public_key.clone(),
+        withdrawal_credentials.clone(),
```
[the remark this answers is not in the document]
Comment (jacobkaufmann) [reply ethereum-consensus/src/phase0/block_processing.rs:406]: yea, looks like we could take from `deposit_message`, but ultimately `get_validator_from_deposit` needs ownership, which is why I went with owned value in `add_validator_to_registry`.

[the remark this answers is not in the document]
Comment (ralexstokes) [reply ethereum-consensus/src/phase0/block_processing.rs:406]: yeah, which is why we should be able to take them away from `deposit_message`

generally I pass by ref until the "leaf" of the AST but passing by value is fine here

File: ethereum-consensus/src/phase0/block_processing.rs [hunk 5]
```diff
@@ -290,16 +292,20 @@ pub fn process_attestation<
     Ok(())
 }
 
-pub fn get_validator_from_deposit(deposit: &Deposit, context: &Context) -> Validator {
-    let amount = deposit.data.amount;
+pub fn get_validator_from_deposit(
+    public_key: BlsPublicKey,
+    withdrawal_credentials: Bytes32,
+    amount: u64,
```
Comment (ralexstokes) [inline ethereum-consensus/src/phase0/block_processing.rs:298]: also `Gwei`?

File: ethereum-consensus/src/phase0/block_processing.rs [hunk 6]
```diff
@@ -309,7 +315,96 @@ pub fn get_validator_from_deposit(deposit: &Deposit, context: &Context) -> Valid
     }
 }
 
-pub(crate) const DEPOSIT_MERKLE_DEPTH: usize = DEPOSIT_CONTRACT_TREE_DEPTH + 1;
+pub const DEPOSIT_MERKLE_DEPTH: usize = DEPOSIT_CONTRACT_TREE_DEPTH + 1;
```
Comment (ralexstokes) [inline ethereum-consensus/src/phase0/block_processing.rs:318]: is there a reason visibility changed here?

**Review by ralexstokes — APPROVED** [review]
looks great! two questions

File: ethereum-consensus/src/phase0/block_processing.rs (hunk 6 above)
Comment (jacobkaufmann) [reply ethereum-consensus/src/phase0/block_processing.rs:318]: it was not being included in the subsequent `spec` modules without that change.

Comment (ralexstokes) [reply ethereum-consensus/src/phase0/block_processing.rs:318]: lets just get rid of this, https://github.com/ralexstokes/ethereum-consensus/pull/379

**Review by ralexstokes — APPROVED** [review]
looks great!

**Pull request merged**